### PR TITLE
Fix BIP39 entropy strength validation

### PR DESCRIPTION
**Pull request by gravityblast**
This PR fixes the way we validate entropy strength value in BIP39.

From:

`if strength%32 > 0 && strength < 128 && strength > 256 {`

To:

`if strength%32 > 0 || strength < 128 || strength > 256 {`

It shouldn't change anything in the core application since we call `MnemonicPhrase` with an hardcoded value from the `account` pkg.

It also fixes the pkg names in tests, and instead of importing the pkg as an external one,
it just tests itself.

### CODE REVIEW

File: extkeys/mnemonic_test.go [hunk 1]
```diff
@@ -21,13 +19,22 @@ type Vector struct {
...
 
-	mnemonic := extkeys.NewMnemonic(extkeys.Salt)
+	mnemonic := NewMnemonic(Salt)
+
+	// test strength validation
+	strengths := []int{127, 129, 257}
+	for _, s := range strengths {
+		_, err := mnemonic.MnemonicPhrase(s, EnglishLanguage)
+		if err != ErrInvalidEntropyStrength {
+			t.Errorf("Entropy strength `%d` should be invalid", s)
```
Comment (adambabik) [inline extkeys/mnemonic_test.go]: Maybe just `'%d'`. Using ` is strange a bit in regular strings.

**Review by adambabik — APPROVED** [review]
LGTM

Can you please post a report with coverage stats for this package (html file ideally)?

**Comment by gravityblast** [comment]
@adambabik I attached an html file. You can also see here the coverage:

```
github.com/status-im/status-go/extkeys/hdkey.go:109:	NewMaster		87.5%
github.com/status-im/status-go/extkeys/hdkey.go:146:	Child			90.3%
github.com/status-im/status-go/extkeys/hdkey.go:214:	BIP44Child		87.5%
github.com/status-im/status-go/extkeys/hdkey.go:239:	Derive			85.7%
github.com/status-im/status-go/extkeys/hdkey.go:254:	Neuter			100.0%
github.com/status-im/status-go/extkeys/hdkey.go:280:	String			94.4%
github.com/status-im/status-go/extkeys/hdkey.go:318:	pubKeyBytes		100.0%
github.com/status-im/status-go/extkeys/hdkey.go:330:	ToECDSA			0.0%
github.com/status-im/status-go/extkeys/hdkey.go:337:	NewKeyFromString	92.3%
github.com/status-im/status-go/extkeys/mnemonic.go:82:	NewMnemonic		91.7%
github.com/status-im/status-go/extkeys/mnemonic.go:104:	AvailableLanguages	100.0%
github.com/status-im/status-go/extkeys/mnemonic.go:121:	MnemonicSeed		100.0%
github.com/status-im/status-go/extkeys/mnemonic.go:126:	MnemonicPhrase		92.9%
github.com/status-im/status-go/extkeys/mnemonic.go:199:	ValidMnemonic		72.7%
github.com/status-im/status-go/extkeys/mnemonic.go:227:	WordList		66.7%
github.com/status-im/status-go/extkeys/mnemonic.go:234:	contains		91.7%
github.com/status-im/status-go/extkeys/mnemonic.go:261:	padByteSlice		100.0%
github.com/status-im/status-go/extkeys/utils.go:17:	splitHMAC		80.0%
github.com/status-im/status-go/extkeys/utils.go:47:	paddedAppend		100.0%
total:							(statements)		89.3%
```

I think we can add more  tests, especially on possible returned errors. I just fixed the BIP39 but I can continue with the rest in another PR just to keep them small, what do you think?

[coverage.html.zip](https://github.com/status-im/status-go/files/1806764/coverage.html.zip)

File: extkeys/mnemonic.go [hunk 2]
```diff
@@ -128,8 +133,8 @@ func (m *Mnemonic) MnemonicPhrase(strength, language Language) (string, error) {
 	// With more entropy security is improved but the sentence length increases.
 	// We refer to the initial entropy length as ENT. The recommended size of ENT is 128-256 bits.
 
-	if strength%32 > 0 && strength < 128 && strength > 256 {
-		return "", errors.New("The mnemonic must encode entropy in a multiple of 32 bits, The recommended size of ENT is 128-256 bits")
+	if strength%32 > 0 || strength < 128 || strength > 256 {
```
Comment (mandrigin) [inline extkeys/mnemonic.go:147]: Can we just make a strength an enum?

Then it would be much harder to make this kind of mistake, harder to misuse APIs.
There aren't that many values.
```go
type Strength  int

const (
         Strength128 = Strength(128)
         ...
)
```

**Review by mandrigin — APPROVED** [review]
Looks good!

A tiny suggestion, but not critical though.

**Comment by adambabik** [comment]
@pilu Thanks!

1. We can add an additional test to make sure a default salt is used when given empty:
```go
func NewMnemonic(salt string) *Mnemonic {
        if len(salt) == 0 {
                salt = Salt
        }
```
2. Check if error is returned if invalid language is selected:
```go
func (m *Mnemonic) WordList(language Language) (*WordList, error) {
        if m.wordLists[language] == nil {
                return nil, fmt.Errorf("language word list is missing (language id: %d)", language)
        }
```
3. That an empty extended key is returned:
```go
func NewKeyFromString(key string) (*ExtendedKey, error) {
        if key == EmptyExtendedKeyString || len(key) == 0 {
                return &ExtendedKey{}, nil
        }
```

Other than that, it looks well covered.

**Comment by mandrigin** [comment]
@adambabik I personally don't really like those default cases, especially in key storage, like
```go
if len(salt) == 0 {
                salt = Salt
}
```

I would make 2 methods instead `NewMnemonicDefaultSalt()` and `NewMnemomic(...)` and panic or return an error if in the second one there isn't any string. That will catch misuses easily.

The same for 
```go
        if key == EmptyExtendedKeyString || len(key) == 0 {
                return &ExtendedKey{}, nil
        }
```

what if `len(key) == 0` is a misuse of the API? would it compromise the key?

File: extkeys/mnemonic.go (hunk 2 above)
Comment (gravityblast) [reply extkeys/mnemonic.go:147]: I like it. Also, if we don't export `Strength`, but we just use `type Strength int`, it would be impossible to call the API with another value like `Strength(123)`. 

Does it make sense @mandrigin ?

Comment (pilu) [reply extkeys/mnemonic.go]: ?

Comment (mandrigin) [reply extkeys/mnemonic.go:147]: absolutely, we shouldn't export what we don't need to.

Comment (gravityblast) [reply extkeys/mnemonic.go:147]: actually I thought go behaves in a different way.
Even if we have:

```
type strength int

const S1 = strength(128)

func foo(s strength) {}
```

You can still call `foo` with another number: `foo(111)`.

So in the end it wouldn't change so much, but maybe is it still more readable? @mandrigin

Comment (mandrigin) [reply extkeys/mnemonic.go:147]: Yeah, it doesn't reduce the need for a check, but it makes the API a bit harder to misuse.
To make is stronger, you can, of course make 
```go
type strength struct {
value int
}
```

but I think that's an overkill.

**Comment by adambabik** [comment]
@mandrigin I don't see any usage which calls `NewMnemonic()` without salt so we can throw an error in such a case to make the salt explicit.

The second question is more spec related I guess, however, I don't see how an empty `ExtendedKey` can be useful, though.

**Comment by mandrigin** [comment]
@adambabik are there any risks if we occasionally (due to a heap corruption or something) get `len(key) == 0` there?

**Comment by adambabik** [comment]
@mandrigin No idea but also I don't see how it can be dangerous here. If there is a way to exploit a memory corruption bug, there are more interesting things to obtain.

Btw. I found another implementation that allows an empty key: https://github.com/btcsuite/btcutil/blob/master/hdkeychain/extendedkey.go#L397-L399

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/btcsuite/btcutil">btcsuite/btcutil</a></strong></div><div>btcutil - Provides bitcoin-specific convenience functions and types</div></blockquote>

**Comment by gravityblast** [comment]
@adambabik @mandrigin I added some more commits following your discussions guys, let me know if you have any other thought please.

File: extkeys/mnemonic.go (hunk 2 above)
Comment (adriacidre) [reply extkeys/mnemonic.go:147]: Hey @pilu would be nice to have this conditional as a method itself, allowing a readable code:

```
	// The mnemonic must encode entropy in a multiple of 32 bits.
	// With more entropy security is improved but the sentence length increases.
	// We refer to the initial entropy length as ENT. The recommended size of ENT is 128-256 bits.
        func (m *Mnemonic) isValidInitialEntropy(strength entropyStrength) bool {
	    return (strength%32 > 0 || strength < 128 || strength > 256)
	}
```
Additionally you'll have an isolated method to run simple unit tests against it :)

Comment (gravityblast) [reply extkeys/mnemonic.go:147]: @adriacidre yes I think I'll try to do another PR for the refactoring. For now I just wanted to fix a "bug" and add tests where needed but I didn't want to change too much code.

**Review by mandrigin — APPROVED** [review]
LGTM

File: extkeys/mnemonic.go [hunk 3]
```diff
@@ -41,6 +41,21 @@ const (
 	totalAvailableLanguages
 )
 
+type entropyStrength int
+
+const (
```
Comment (adambabik) [inline extkeys/mnemonic.go]: `iota` FTW!

```go
const (
    EntropyStrength128 entropyStrength = 128 + (32 * iota)
    EntropyStrength160
    EntropyStrength192
    EntropyStrength224
    EntropyStrength256
)
```

File: extkeys/mnemonic.go [hunk 4]
```diff
@@ -220,7 +240,7 @@ func (m *Mnemonic) ValidMnemonic(mnemonic string, language Language) bool {
 
 // WordList returns list of words for a given language
 func (m *Mnemonic) WordList(language Language) (*WordList, error) {
-	if m.wordLists[language] == nil {
+	if int(language) < 0 || int(language) > len(m.wordLists)-1 || m.wordLists[language] == nil {
```
Comment (adambabik) [inline extkeys/mnemonic.go:239]: The previous code looked fine. I would probably write it like this:
```go
words, ok := m.wordLists[language]
if !ok {
  return nil, fmt.Errorf("language word list is missing (language id: %d)", language)
}

return words
```

but the previous version should work the same in this case.

File: extkeys/mnemonic_test.go [hunk 5]
```diff
@@ -18,16 +16,54 @@ type Vector struct {
...
+
+	customSalt := "custom-salt"
+	m2 := NewMnemonic(customSalt)
+	if m2.salt != customSalt {
+		t.Errorf("expected %q, got: %q", customSalt, m2.salt)
+	}
+}
+
+func TestMnemonic_WordList(t *testing.T) {
+	m := NewMnemonic("")
```
Comment (adambabik) [inline extkeys/mnemonic_test.go]: You can use default salt (`NewMnemonic(Salt)`) if you test a regular case. Let's the most explicit usage in tests.

**Review by adambabik — CHANGES_REQUESTED** [review]

File: extkeys/mnemonic.go (hunk 3 above)
Comment (gravityblast) [reply extkeys/mnemonic.go]: much better

File: extkeys/mnemonic.go (hunk 4 above)
Comment (gravityblast) [reply extkeys/mnemonic.go:239]: @adambabik I think `words, ok := m.wordLists[language]` would work with a map but in this case it's an array: `wordLists [totalAvailableLanguages]*WordList`, is that correct?

Comment (adambabik) [reply extkeys/mnemonic.go:239]: Oh, shoot, I thought it's a map. Let's keep your code then! 

PS we can write `int(language) >= len(m.wordLists)`.

**Review by adambabik — APPROVED** [review]
